docs/architecture/adr-070-hasher.md

```diff
@@ -0,0 +1,146 @@
+# ADR 070: hasher 
+
+## Changelog
+
+- 26 July 2021: Initial Draft (@jayt106)
+
+## Status
+
+Proposed
+
+## Context
+
+The current hashing algorithm in the Tendermint core project relies on the Go standard package [sha256](https://pkg.go.dev/crypto/sha256) and the implementation [crypto/tmhash](https://github.com/tendermint/tendermint/blob/master/crypto/tmhash/hash.go) only. The user might want to extend the hash algorithm for the different project's
+requirements. For example, the [ethermint](https://github.com/tharsis/ethermint) tries to bring the transaction to the Ethereum network but the transaction hash in the Ethereum world represents it in a different way [RLP hash](https://github.com/ethereum/go-ethereum/blob/92b8f28df3255c6cef9605063850d77b46146763/core/types/transaction.go#L368).
+
+Alternatively, some advanced hash algorithm provides faster-hashing speed, it might increase some scabilities to the project, like the mempool processing, trie proccessing, database data store/retrieve, and so on. For example, the supplementary Go cryptography libraries implemented [blake2b](https://pkg.go.dev/golang.org/x/crypto/blake2b) has roughly 2.25X faster than the `sha256` we mentioned above (both testing under the AVX2 supported CPU processor) [ref](https://github.com/SimonWaldherr/golang-benchmarks#hash). The [sha256-simd](https://github.com/minio/sha256-simd/) provides up to 8x faster than the `crypto/sha256` if the CPU processor supports [AVX512](https://en.wikipedia.org/wiki/AVX-512). The [xxHash](https://github.com/cespare/xxhash) has been using in [Polkadot](https://substrate.dev/docs/en/knowledgebase/advanced/cryptography#hashing-algorithms) project and it provides dramatically speed hashing can be using certain scenarios like the hash of immutable data input.
+
+## Decision
+TBD
+
+## Detailed Design
+
+### The new hashing struct
+Introduce `Hasher` interface to inherit [hash.Hash](https://pkg.go.dev/hash#Hash) and extend it with the current Tendermint core existing functions. The `Hasher` implementation is required in each supporting hash algorithm.
+
+```go
+package tmhash
+
+type Hasher interface {
+    hash.Hash
+    SumTruncated(bz []byte) []byte
```

Comment:
why do we need `SumTruncated`? Sum should work - and should be customized by a developer if needed. Eg if we want to truncate a hash function which produces 48 bytes, then we can wrap it with a new hasher. So I wouldn't introduce a new interface.

Reply:
@robert-zaremba, thanks for the comments. I didn't notice the discussion in SDK. I will check these contexts.

If was a function in `tmhash` package. I agreed it can be separated from the interface. Currently only used in the pubkey to address convertion

Reply:
Why not using the `hash.Hash.Sum` function? if we want to customize we will create a wrapper.

Reply:
I would avoid defining a new interface. Let's stick with the stdlib `Hash` interface if possible.